master/pkg/searcher/asha.go

```diff
@@ -0,0 +1,244 @@
+		len(r.metrics),
+		func(i int) bool { return r.metrics[i].metric > metric },
+	)
+	promoteNow := insertIndex < numPromote
+
+	r.metrics = append(r.metrics, trialMetric{})
+	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
+	r.metrics[insertIndex] = trialMetric{
+		requestID: requestID,
+		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
+	}
+
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
+		return []RequestID{requestID}
+	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
+	}
+
+	return nil
+}
+
+func (s *asyncHalvingSearch) initialOperations(ctx context) ([]Operation, error) {
+	// The number of initialOperations will control the degree of parallelism
+	// of the search experiment since we guarantee that each validationComplete
+	// call will return a new train workload until we reach MaxTrials.
+
+	// We will use searcher config field if available.
+	// Otherwise we will default to a number of trials that will
+	// guarantee at least one trial at the top rung.
+	var ops []Operation
+	var maxConcurrentTrials int
+
+	if s.MaxConcurrentTrials != nil {
+		maxConcurrentTrials = min(*s.MaxConcurrentTrials, s.MaxTrials)
+	} else {
+		maxConcurrentTrials = max(
+			min(int(math.Pow(s.Divisor, float64(s.NumRungs-1))), s.MaxTrials),
+			1)
+	}
+
+	for trial := 0; trial < maxConcurrentTrials; trial++ {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		ops = append(ops, create)
+		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
+	}
+	return ops, nil
+}
+
+func (s *asyncHalvingSearch) trainCompleted(
+	ctx context, requestID RequestID, message Workload,
+) ([]Operation, error) {
+	s.totalStepsCompleted++
+	return nil, nil
+}
+
+func (s *asyncHalvingSearch) validationCompleted(
+	ctx context, requestID RequestID, message Workload, metrics ValidationMetrics,
+) ([]Operation, error) {
+	// Extract the relevant metric as a float.
+	metric, err := metrics.Metric(s.AsyncHalvingConfig.Metric)
+	if err != nil {
+		return nil, err
+	}
+	if !s.AsyncHalvingConfig.SmallerIsBetter {
+		metric *= -1
+	}
+
+	return s.promoteAsync(ctx, requestID, message, metric), nil
+}
+
+func (s *asyncHalvingSearch) promoteAsync(
+	ctx context, requestID RequestID, message Workload, metric float64,
+) []Operation {
+	// Upon a trial is finished, we should return at least one more trial to train unless
+	// the bracket of successive halving is finished.
+	rungIndex := s.trialRungs[requestID]
+	rung := s.rungs[rungIndex]
+	rung.outstandingTrials--
+	addedNewTrial := false
```

Comment:
blocking: can we rename this to `promotedTrial` or something? Unless we actually mean that we added a new trial to the experiment rather than promoting an existing trial?

Reply:
outstandingTrials is used to track pending trials not just promoted ones.  we increment outstanding trials when adding a new trial to the bottom rung and when promoting trials to a higher rung.

Reply:
I am referring to `addedNewTrial`

Reply:
ok, i changed it to addedTrainWorkload